ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_sender_HPerf_LPrec.cpp

```diff
@@ -0,0 +1,100 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdint>
+#include "tt_metal/hw/inc/dataflow_api.h"
+#include "debug/assert.h"
+#include "debug/dprint.h"
+#include <array>
+#include "tt_metal/hw/inc/ethernet/dataflow_api.h"
+
+
+struct addr_sem_pair {
+    uint32_t addr;
+    uint32_t sem_addr;
+};
+
+
+static constexpr bool DISABLE_CONTEXT_SWITCHING = true;
+static constexpr uint8_t NUM_CHANNELS = 8;
+
+FORCE_INLINE void eth_setup_handshake(uint32_t handshake_register_address) {
+    eth_send_bytes(handshake_register_address, handshake_register_address, 16);
+    eth_wait_for_receiver_done();
+}
+
+FORCE_INLINE
+void eth_send_bytes_over_channel_payload_only(
+    uint32_t src_addr,
+    uint32_t dst_addr,
+    uint32_t num_bytes,
+    volatile eth_channel_sync_t *channel_sync) {
+    channel_sync->bytes_sent = num_bytes;
+    channel_sync->receiver_ack = 0;
+    internal_::eth_send_packet(
+        0, src_addr >> 4, dst_addr >> 4, num_bytes >> 4);
+}
+
+FORCE_INLINE void perform_noc_read(
+    uint32_t channels_sem_addrs, uint32_t ring_index)
+{
+    volatile uint32_t *sem_addr = reinterpret_cast<volatile uint32_t*>(channels_sem_addrs);
+    eth_noc_semaphore_wait(sem_addr, 1);
+    *sem_addr = 0;
+}
+
+FORCE_INLINE void perform_eth_write(
+    uint32_t &channels_addrs,
+    volatile eth_channel_sync_t* eth_channel_syncs,
+    uint32_t page_size,
+    uint32_t ring_index)
+{
+    uint32_t buffer_addr = channels_addrs;
+    eth_send_bytes_over_channel_payload_only(buffer_addr, buffer_addr, page_size + sizeof(eth_channel_sync_t), eth_channel_syncs);
```

Comment:
loop while `!eth_txq_is_busy()` before calling (this goes with any eth call). Wait for some time before context switching.